web/src/app/util/QuerySetFavoriteButton.js

```diff
@@ -0,0 +1,92 @@
+import React from 'react'
+import p from 'prop-types'
+import gql from 'graphql-tag'
+import { graphql2Client } from '../apollo'
+import Query from '../util/Query'
+import { Mutation } from 'react-apollo'
+import { SetFavoriteButton } from './SetFavoriteButton'
+import { oneOfShape } from '../util/propTypes'
+
+const queries = {
+  service: gql`
+    query serviceFavQuery($id: ID!) {
+      service(id: $id) {
+        id
+        isFavorite
+      }
+    }
+  `,
+  rotation: gql`
+    query rotationFavQuery($id: ID!) {
+      rotation(id: $id) {
+        id
+        isFavorite
+      }
+    }
+  `,
+}
+
+const mutation = gql`
+  mutation setFav($input: SetFavoriteInput!) {
+    setFavorite(input: $input)
+  }
+`
+
+export function QuerySetFavoriteButton(props) {
+  let typeName = ''
+  let id = ''
+  if (props.rotationID) {
+    typeName = 'rotation'
+    id = props.rotationID
+  } else if (props.serviceID) {
+    typeName = 'service'
+    id = props.serviceID
+  }
```

Comment:
if id/typeName isn't one of `serviceID` or `rotationID` it will throw a proptype error, but it may also cause the page to crash because of an apollo error in the query and variables. we should default to returning `null` if either of these isn't set